include/rmm/mr/host/host_memory_resource.hpp

```diff
@@ -0,0 +1,168 @@
+ *
+ * This is based on `std::pmr::memory_resource`:
+ * https://en.cppreference.com/w/cpp/memory/memory_resource
+ *
+ * When C++17 is available for use in RMM, `rmm::host_memory_resource` should
+ * inherit from `std::pmr::memory_resource`.
+ *
+ * This class serves as the interface that all host device memory resource
+ * implementations must satisfy.
+ *
+ * There are two private, pure virtual functions that all derived classes must
+ * implement: `do_allocate` and `do_deallocate`. Optionally, derived classes may
+ * also override `is_equal`. By default, `is_equal` simply performs an identity
+ * comparison.
+ *
+ * The public, non-virtual functions `allocate`, `deallocate`, and `is_equal`
+ * simply call the private virtual functions. The reason for this is to allow
+ * implementing shared, default behavior in the base class. For example, the
+ * base class' `allocate` function may log every allocation, no matter what
+ * derived class implementation is used.
+ *
+ *---------------------------------------------------------------------------**/
+class host_memory_resource {
+ public:
+  host_memory_resource() = default;
+  virtual ~host_memory_resource() = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates memory on the host of size at least `bytes` bytes.
+   *
+   * The returned storage is aligned to the specified `alignment` if supported,
+   * and to `alignof(std::max_align_t)` otherwise.
+   *
+   * @throws std::bad_alloc When the requested `bytes` and `alignment` cannot be
+   * allocated.
+   *
+   * @param bytes The size of the allocation
+   * @param alignment Alignment of the allocation
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  void* allocate(std::size_t bytes,
+                 std::size_t alignment = alignof(std::max_align_t)) {
+    return do_allocate(bytes, alignment);
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * `p` must have been returned by a prior call to `allocate(bytes,alignment)`
+   * on a `host_memory_resource` that compares equal to `*this`, and the storage
+   * it points to must not yet have been deallocated, otherwise behavior is
+   * undefined.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param alignment Alignment of the allocation. This must be equal to the
+   *value of `alignment` that was passed to the `allocate` call that returned
+   *`p`.
+   * @param stream Stream on which to perform deallocation
+   *---------------------------------------------------------------------------**/
+  void deallocate(void* p, std::size_t bytes,
```

Comment:
Should this function be `noexcept`?

Reply:
I'd thought so too, but for whatever reason neither [`std::pmr::memory_resource::deallocate`](https://en.cppreference.com/w/cpp/memory/memory_resource/deallocate), [`do_deallocate`](https://en.cppreference.com/w/cpp/memory/memory_resource/do_deallocate), nor the concrete derived types mark their [`deallocate`](https://en.cppreference.com/w/cpp/memory/synchronized_pool_resource/do_deallocate) functions with `noexcept`. 

I'm not really sure why.

Reply:
I learned the answer is due to the ["Lakos Rule" ](https://quuxplusone.github.io/blog/2018/04/25/the-lakos-rule/)

> Remove noexcept specifications from each library function having a narrow contract, typically (but not always) indicated by the presence of a Requirements: clause.

Basically, since `deallocate` has undefined behavior if anything other than a pointer returned from `allocate` is passed in (a requirement), that means it shouldn't be marked `noexcept`.